server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoClusterHistogramAggregationBuilder.java

```diff
@@ -0,0 +1,159 @@
+    }
+
+    public AutoClusterHistogramAggregationBuilder setNumBuckets(int numBuckets){
+        if (numBuckets <= 0) {
+            throw new IllegalArgumentException(NUM_BUCKETS_FIELD.getPreferredName() + " must be greater than 0 for ["
+                + name + "]");
+        }
+        this.numBuckets = numBuckets;
+        return this;
+    }
+
+    public AutoClusterHistogramAggregationBuilder setShardSize(int shardSize){
+        if (shardSize < numBuckets) {
+            throw new IllegalArgumentException(SHARD_SIZE_FIELD.getPreferredName() + " must be greater than or equal to "
+                + NUM_BUCKETS_FIELD.getPreferredName() + "for [" + name + "]");
+        }
+        this.shardSize = shardSize;
+        return this;
+    }
+
+    public int getNumBuckets(){ return numBuckets; }
+
+    public int getShardSize(){ return shardSize; }
+
+    public static AutoClusterHistogramAggregationBuilder parse(String aggregationName,
+                                                               XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new AutoClusterHistogramAggregationBuilder(aggregationName), null);
+    }
+
+    /** Create a new builder with the given name. */
+    public AutoClusterHistogramAggregationBuilder(String name) {
+        super(name, ValuesSourceType.NUMERIC, ValueType.DOUBLE); // Only clustering double fields for now
+    }
+
+    /** Read in object data from a stream, for internal use only. */
+    public AutoClusterHistogramAggregationBuilder(StreamInput in) throws IOException {
+        super(in, ValuesSourceType.NUMERIC, ValueType.DATE);
+        numBuckets = in.readVInt();
+    }
+
+    @Override
+    protected AggregationBuilder shallowCopy(AggregatorFactories.Builder factoriesBuilder, Map<String, Object> metaData) {
+        return new AutoClusterHistogramAggregationBuilder(this, factoriesBuilder, metaData);
+    }
+
+    protected AutoClusterHistogramAggregationBuilder(AutoClusterHistogramAggregationBuilder clone,
+                                                     AggregatorFactories.Builder factoriesBuilder,
+                                                     Map<String, Object> metaData) {
+        super(clone, factoriesBuilder, metaData);
+        this.numBuckets = clone.numBuckets;
+    }
+
+
+    @Override
+    protected void innerWriteTo(StreamOutput out) throws IOException {
+        out.writeVInt(numBuckets);
+    }
+
+    // Produce a factory given the members in this builder
+    @Override
+    protected ValuesSourceAggregatorFactory<ValuesSource.Numeric, ?> innerBuild(
+        SearchContext context,
+        ValuesSourceConfig<ValuesSource.Numeric> config, AggregatorFactory<?> parent,
+        AggregatorFactories.Builder subFactoriesBuilder) throws IOException {
+
+        Settings settings = context.getQueryShardContext().getIndexSettings().getNodeSettings();
+        int maxBuckets = MultiBucketConsumerService.MAX_BUCKET_SETTING.get(settings);
```

Comment:
Each shard should should individually return more than the `numBuckets` buckets. Then, the agglomerative clustering will reduce them to exactly `numBuckets` buckets.

If shards are returning less buckets, this can be fixed by increasing the `shard_size` and `cache_limit` parameters.